src/components/Context.js

```diff
@@ -183,49 +219,51 @@ class ProductProvider extends Component {
   addTotals = () => {
     let subTotal = 0;
     this.state.cart.map((item) => (subTotal += item.total));
-    const tempTax = subTotal * 0.1;
-    const tax = parseFloat(tempTax.toFixed(2));
-    const total = subTotal + tax;
+    const total = subTotal;
+
     this.setState(() => {
       return {
         cartSubTotal: subTotal,
-        cartTax: tax,
         cartTotal: total,
       };
     });
+    this.addPromotion();
   };
 
-  addPromotionToCart = (couponCode) => {
-    client.vouchers.get(couponCode, (error, result) => {
-      if (error) {
-        toast.error('Promotion not found');
-        return error;
-      }
-      const voucher = result;
-      console.log(voucher);
+  addPromotionToCart = async (couponCode) => {
+    try {
+      const voucher = await new Promise((resolve, reject) => {
+        window.Voucherify.validate(couponCode, (response) => {
+          if (response.valid) {
+            resolve(response);
+          } else {
+            reject(new Error(response.reason));
+          }
+        });
+      });
+
       this.setState(
```

Comment:
1. let's change it to a `dispatch(ADD_COUPON, { pass the voucher here })`

2. go back to your reducer function - let's extract a reusable function which does all the discount calculation:
```
const calculateCart = (cartTotal, cartDiscount) => {
  // TODO calculate it
  return {
    cartTotalAfterPromotion: 123,
    discounted: 234,
  };
}
```

3. in the reducer, let's use this function for ADD_PROMOTION

4. use it also for `ADD_COUPON` - the new switch/case in your reducer

This way you don't need to call this.addPromotion() in line:254